### PR TITLE
MC-13864: Consumer always read config from memory

### PR DESCRIPTION
**Pull request by vzabaznov**
## Problem

Right now our consumer depends on config/objects in-memory state. After application was bootstrapped it initialize a lot of object that are always kept in memory after first load - config, websites, etc.., since our consumers are long-living and their live time limited only by message count it leads to wrong im-memory config/objects states and we need to introduce the way to reinit them
## Solution

Introduce poison pill - some message stored in Data Storage that signals that consumer must be reinited

## Requested Reviewers

@kandy @paliarush @vrann

### CODE REVIEW

File: design-documents/asynchronous-operations/poison-pill.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+Before process new message consumer should ask Data Storage if there are poison pill for him. If yes, consumer dies and next cron:run will start new instance of consumer.
+
+#### Acceptance Criteria Fulfillment
+
+1. If consumer takes message to process after poison pill was placed, consumer must be reinited
+
+#### Extension Points and Scenarios
+
+We will put poison pill into Data Storage based on common Magento extension points like events and plugins
```
Comment (joni-jones) [inline design-documents/asynchronous-operations/poison-pill.md]: As our goal is modules decoupling, we should not add new dependencies between modules via plugins or observers.

File: design-documents/asynchronous-operations/poison-pill.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+## Glossary
+
+poison pill
+: message in Data Storage that signals that consumer should die
```
Comment (joni-jones) [inline design-documents/asynchronous-operations/poison-pill.md]: What kind of consumer do you mean? A message queue, cron?

File: design-documents/asynchronous-operations/poison-pill.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+### Overview
+
+Right now our consumer depends on config/objects in-memory state. After application was bootstrapped it initialize a lot of object that are always kept in memory after first load - config, websites, etc.., since our consumers are long-living and their live time limited only by message count it leads to wrong im-memory config/objects states and we need to introduce the way to reinit them.
+
+### Design
+
+Before process new message consumer should ask Data Storage if there are poison pill for him. If yes, consumer dies and next cron:run will start new instance of consumer.
```
Comment (joni-jones) [inline design-documents/asynchronous-operations/poison-pill.md]: How a consumer will communicate with data storage?

File: design-documents/asynchronous-operations/poison-pill.md (hunk 1 above)
Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: Basically, we are not going to use in-process communication

File: design-documents/asynchronous-operations/poison-pill.md (hunk 3 above)
Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: I think yeas, we are going to introduce the basic implementation inside our framework and other modules will use it

File: design-documents/asynchronous-operations/poison-pill.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+## Glossary
+
+poison pill
+: message in Data Storage that signals that consumer should die
+
+### Overview
+
+Right now our consumer depends on config/objects in-memory state. After application was bootstrapped it initialize a lot of object that are always kept in memory after first load - config, websites, etc.., since our consumers are long-living and their live time limited only by message count it leads to wrong im-memory config/objects states and we need to introduce the way to reinit them.
```
Comment (buskamuza) [inline design-documents/asynchronous-operations/poison-pill.md]: Could you clarify what "wrong" means? Is it a state modified by the processes, so it is changed?

Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: wrongs means that if someone will change the state, consumer will running with outdated in cache, yeah outdated is more accurate meaning right here

File: design-documents/asynchronous-operations/poison-pill.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: The message will be added after some event is triggered, e.g. website was created

File: design-documents/asynchronous-operations/poison-pill.md (hunk 3 above)
Comment (joni-jones) [reply design-documents/asynchronous-operations/poison-pill.md]: Could you add more details about API which should be introduced?

File: design-documents/asynchronous-operations/poison-pill.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: Adding variable on cloud triggers redeploy, which will restart consumers by default

File: design-documents/asynchronous-operations/poison-pill.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+poison pill
+: message in Data Storage that signals that consumer should die
+
+### Overview
+
+Right now our consumer depends on config/objects in-memory state. After application was bootstrapped it initialize a lot of object that are always kept in memory after first load - config, websites, etc.., since our consumers are long-living and their live time limited only by message count it leads to outdated im-memory config/objects states and we need to introduce the way to reinit them.
+
+### Design
+
+Before process new message consumer should ask Data Storage if there are poison pill for him. If yes, consumer dies and next cron:run will start new instance of consumer.
```
Comment (igrybkov) [inline design-documents/asynchronous-operations/poison-pill.md]: If there is a possibility that several consumers would be started simultaneously, how are you going to deliver your poison pill for each of them?

Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: Each of them will compare version of poison pill in his memory vs latest in DB

File: design-documents/asynchronous-operations/poison-pill.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Put poison pill inside DB.
+ *
+ * @throws \Exception
+ * @return int
+ */
+public function put();
+```
+
+Before process new message, consumer will compare version of PoisonPill in-memory state with latest in DB. If version is different consumer dies and next cron:run will start new instance of consumer.
```
Comment (paliarush) [inline design-documents/asynchronous-operations/poison-pill.md]: Does it mean that there will be at least one MySQL DB query for each consumed message?

Comment (vzabaznov) [reply design-documents/asynchronous-operations/poison-pill.md]: Yes

Comment (paliarush) [reply design-documents/asynchronous-operations/poison-pill.md]: Please add warning to the design document for the team implementing this feature to verify performance implications.

Comment (tariqjawed83) [reply design-documents/asynchronous-operations/poison-pill.md:27]: @paliarush @vzabaznov I am concerned about this point specifically, if for each read we are hitting the database query, then what is the purpose of storing configurations in-memory? we might lose the benefits of in-memory optimizations. We should think differently here, on invalidating the cache on event based.

**Pull request merged**